Contact
=======

github: '@maxemil <ttps://github.com/maxemil>'
'maxemil.github.io <https://maxemil.github.io/>'

Citation
========
If you use PhyloMagnet, please cite the article:
Max E Sch√∂n, Laura Eme, Thijs J G Ettema, PhyloMagnet: fast and accurate screening of short-read meta-omics data using gene-centric phylogenetics, Bioinformatics, btz799, https://doi.org/10.1093/bioinformatics/btz799

Also cite all the software that PhyloMagnet uses internally, depending on your choice of tree and alignment software:

diamond: doi:10.1038/nmeth.3176 and https://github.com/bbuchfink/diamond

megan: doi:10.1371/journal.pcbi.1004957 and doi:10.1186/s40168-017-0233-2 and http://ab.inf.uni-tuebingen.de/software/megan6/

papara: doi:10.1093/bioinformatics/btr320 and https://cme.h-its.org/exelixis/web/software/papara/index.html

epa-ng: doi:10.1093/sysbio/syy054 and https://github.com/Pbdas/epa-ng

gappa: doi:10.1101/647958 and https://github.com/lczech/gappa


iqtree: doi:10.1093/molbev/msu300 and http://www.iqtree.org/

FastTree: doi:10.1371/journal.pone.0009490 and http://www.microbesonline.org/fasttree/

raxml-ng: doi:10.1093/bioinformatics/btz305 and https://github.com/amkozlov/raxml-ng


mafft: doi:10.1093/molbev/mst010 and https://mafft.cbrc.jp/alignment/software/

prank: doi:10.1007/978-1-62703-646-7_10 and http://wasabiapp.org/software/prank/
